### PR TITLE
[LoopPeel] Support min/max intrinsics in loop peeling

### PR DESCRIPTION
**Pull request by skachkov-llvm**
Motivating example: https://godbolt.org/z/qfxz8rKac

This example shows very different loop peeling behaviour for similar code depending on whether icmp+select was canonicalized to min/max intrinsic or not.

This patch adds processing of min/max intrinsics in LoopPeel in the similar way as it was done for conditional statements: for min/max(IterVal, BoundVal) we peel iterations where IterVal < BoundVal for monotonically increasing IterVal; for monotonically decreasing IterVal we peel iterations where IterVal > BoundVal (strict comparision predicates are used to minimize number of peeled iterations).

Compile-time results on llvm test-suite (including SPEC benchmarks) shows 3 tests affected by this:
```
Program                                                                         loop-peel.NumPeeled                             
                                                                                baseline            loop-peel-min-max      diff 
test-suite :: MultiSource/Benchmarks/mediabench/mpeg2/mpeg2dec/mpeg2decode.test   0.00                2.00                  inf%
                   test-suite :: MultiSource/Applications/JM/lencod/lencod.test   7.00                9.00                 28.6%
              test-suite :: External/SPEC/CINT2006/464.h264ref/464.h264ref.test   6.00                7.00                 16.7%
```

### CODE REVIEW

**Review by nikic — COMMENTED** [review]
The idea here looks reasonable, but I wonder if we can reuse the existing code more? Extract the part that works on a decomposed icmp into a separate function and then call it with the icmp for the select expansion of the min/max?

**Comment by skachkov-sc** [comment]
> The idea here looks reasonable, but I wonder if we can reuse the existing code more? Extract the part that works on a decomposed icmp into a separate function and then call it with the icmp for the select expansion of the min/max?

The ammount of shared code will be quite small (because for min/max intrinsics we don't need to process very special equality comparision case), but we can try to split part that finds number of iterations where predicate is still known.

**Review by caojoshua — APPROVED** [review]
Nice idea, LGTM. Please wait for review from nikic@ as well

File: llvm/lib/Transforms/Utils/LoopPeel.cpp
Comment (nikic) [inline llvm/lib/Transforms/Utils/LoopPeel.cpp:476]: Don't we also have to check for a monotonic predicate, as the icmp handling does? Otherwise we might pointlessly peel a wrapping addrec?

Comment (skachkov-sc) [reply llvm/lib/Transforms/Utils/LoopPeel.cpp:480]: Thank you, there is actually a problem here; added test case for wrapping AddRec (test_wrap). Instead of using ScalarEvolution::getMonotonicPredicateType(), I'm suggesting to check SCEVAddRec NUW/NSW flags, because at that point we already know that Step is strictly positive or negative, so no need to check this particular condition again.

**Comment by nikic** [comment]
The arm_mult_q15.ll test is failing.

File: llvm/test/Transforms/LoopUnroll/peel-loop-min-max-intrinsics.ll
Comment (nikic) [inline llvm/test/Transforms/LoopUnroll/peel-loop-min-max-intrinsics.ll:15]:
```suggestion
define void @test_umin(i32 %N) {
```
Test names could be a bit more meaningful...

File: llvm/test/Transforms/LoopUnroll/peel-loop-min-max-intrinsics.ll
Comment (nikic) [inline llvm/test/Transforms/LoopUnroll/peel-loop-min-max-intrinsics.ll:13]:
```suggestion
declare void @foo(i32)
declare void @bar(i8)
```
Intrinsic declarations are no longer needed.

File: llvm/test/Transforms/LoopUnroll/peel-loop-min-max-intrinsics.ll
Comment (nikic) [inline llvm/test/Transforms/LoopUnroll/peel-loop-min-max-intrinsics.ll:2]:
```suggestion
; RUN: opt < %s -S -passes=loop-unroll -unroll-peel-max-count=3 | FileCheck %s
```
I think that would show that peeling actually stops after the right number of iterations?

Comment (skachkov-sc) [reply llvm/test/Transforms/LoopUnroll/peel-loop-min-max-intrinsics.ll]: Increased the limit; also added special test (test_max_count_threshold) to check that if we want to peel more than max-peel-count, we shouldn't peel loop at all (because such transform will not help to eliminate min/max intrinsic in loop)

File: llvm/test/Transforms/LoopUnroll/peel-loop-min-max-intrinsics.ll
Comment (skachkov-sc) [reply llvm/test/Transforms/LoopUnroll/peel-loop-min-max-intrinsics.ll]: done

**Comment by skachkov-sc** [comment]
> The arm_mult_q15.ll test is failing. Maybe need to guard against vector min/max?

Yep, didn't have ARM backend enabled in local build... Added check to process only IntegerType min/max

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**
